### ISSUE TITLE
Some permissions are not restored properly

### ISSUE THREAD

**Issue by Cyberavater** [MuntashirAkon/AppManager#388]
-̷ ̷[̷ ̷]̷ ̷I̷f̷ ̷a̷n̷ ̷a̷p̷p̷ ̷w̷a̷s̷ ̷b̷a̷c̷k̷e̷d̷ ̷u̷p̷ ̷w̷i̷t̷h̷ ̷b̷a̷c̷k̷g̷r̷o̷u̷n̷d̷ ̷b̷a̷t̷t̷e̷r̷y̷ ̷u̷s̷a̷g̷e̷ ̷r̷e̷s̷t̷r̷i̷c̷t̷i̷o̷n̷,̷ ̷a̷f̷t̷e̷r̷ ̷r̷e̷s̷t̷o̷r̷i̷n̷g̷ ̷t̷h̷e̷ ̷s̷e̷t̷t̷i̷n̷g̷s̷ ̷i̷s̷ ̷s̷t̷u̷c̷k̷,̷ ̷i̷.̷e̷.̷ ̷c̷a̷n̷'̷t̷ ̷c̷h̷a̷n̷g̷e̷ ̷i̷t̷ ̷a̷n̷y̷m̷o̷r̷e̷.̷ ̷N̷o̷w̷,̷ ̷t̷h̷e̷ ̷s̷e̷t̷t̷i̷n̷g̷ ̷s̷t̷a̷t̷e̷s̷ ̷t̷h̷a̷t̷ ̷i̷t̷'̷s̷ ̷r̷e̷s̷t̷r̷i̷c̷t̷e̷d̷,̷ ̷i̷n̷ ̷r̷e̷a̷l̷i̷t̷y̷,̷ ̷i̷t̷'̷s̷ ̷n̷o̷t̷.̷ - This is actually AOSP's settings page bug and has nothing to do with AM (read bellow to learn further).
- [ ] It seems backing up permission only backs up Allowed ones but doesn't back up the data about "Deny and don't ask again".

**Comment by MuntashirAkon** [comment]
> * [ ]  If an app was backed up with background battery usage restriction, after restoring the settings is stuck, i.e. can't change it anymore. Now, the setting states that it's restricted, in reality, it's not.

This could be a bug in the settings as it doesn't expect any third-party apps to modify the settings (seems to present in all the AOSP-based ROMs, so this bug is present in the AOSP, I believe). If you have disabled battery optimisation, for instance, the settings may only list it after a reboot although the setting is applied successfully. You can check/change battery related settings of the app in question in the app info page. So, you don't have to rely on the buggy implementation by the AOSP.

**Comment by MuntashirAkon** [comment]
> * [ ]  It seems backing up permission only backs up Allowed ones but doesn't back up the data about "Deny and don't ask again".

Yeah, AM currently saves only allowed permissions. This will be changed in future.

**Comment by Cyberavater** [comment]
> This could be a bug in the settings as it doesn't expect any third-party apps to modify the settings (seems to present in all the AOSP-based ROMs, so this bug is present in the AOSP, I believe). If you have disabled battery optimisation, for instance, the settings may only list it after a reboot although the setting is applied successfully. You can check/change battery related settings of the app in question in the app info page. So, you don't have to rely on the buggy implementation by the AOSP.

I've asked a question here: https://android.stackexchange.com/questions/236592/restrict-apps-background-battery-usage-using-terminal-root-granted
Let's see if someone from there can help.

**Comment by MuntashirAkon** [comment]
Yes, it can be changed using terminal. See https://github.com/MuntashirAkon/AppManager/issues/121#issuecomment-739652765.

**Comment by Cyberavater** [comment]
> Yes, it can be changed using terminal. See [#121 (comment)](https://github.com/MuntashirAkon/AppManager/issues/121#issuecomment-739652765).

So, you're suggesting the problem is something else?

![Screenshot_20210422-131101959 (1)](https://user-images.githubusercontent.com/32187927/115671520-69b1fd80-a36c-11eb-9b45-bf1b92816b0c.jpg)

How about giving AM this permission and then try again?

**Comment by MuntashirAkon** [comment]
> So, you're suggesting the problem is something else?

Yes, a bug in the AOSP settings app. The operation itself works as expected.

> How about giving AM this permission and then try again?

Wouldn't work. Battery optimisation setting isn't part of the system settings. You'd need `android.permission.DUMP`.

**Comment by Cyberavater** [comment]
> You can check/change battery related settings of the app in question in the app info page. So, you don't have to rely on the buggy implementation by the AOSP.

Do you mean App in app info page of AM? That only has battery optimization, not the background usage restriction.
![unnamed](https://user-images.githubusercontent.com/32187927/115674916-f4482c00-a36f-11eb-9911-926ac35da540.jpg)

**Comment by MuntashirAkon** [comment]
> Do you mean App in app info page of AM? That only has battery optimization, not the background usage restriction.

Ah, you mean the first option. It's actually based on two app ops: `RUN_IN_BACKGROUND` and `RUN_ANY_IN_BACKGROUND`. You can either revoke them manually from the app ops tab or by the **prevent background operation** option available in the batch operations as well as in the running apps page.

**Comment by Cyberavater** [comment]
> It's actually based on two app ops: RUN_IN_BACKGROUND and RUN_ANY_IN_BACKGROUND. You can either revoke them manually from the app ops tab or by the prevent background operation option available in the batch operations as well as in the running apps page.

Nice! it works perfectly! Now i've deduced the issue here. 
If after restoring _without_ "Extra" component, and then mannually restrict it using AM's App Ops menu, it works perfectly (The setting is not stuck now on AOSP's own settings and works as intended).
If I restore the app **with** the "Extra" component, the automatic script that pushes the switch does it faulty, thus the issue occurs, I think you should check specifically here as I don't know what fault it's actullay making, all I was able to deduce with test that it works faulty.

**Comment by MuntashirAkon** [comment]
Not sure where you're going. What AM does is perfectly correct. It's the settings page that has the bug. Consider the following situations:

**Situation#1**
0. Take a fresh app
1. Disable background restriction and enable battery optimisation in settings page
2. Take backups
3. Reset both settings
4. Restore the backup
5. Check settings page: it should display correctly.

**Situation#2**
0. Take a fresh app
1. Enable both background restriction and battery optimisation in settings page
2. Take backups
3. Reset both settings
4. Restore the backup
5. Check settings page: it should display correctly.

**Situation#3
0. Take a fresh app
1. Disable both background restriction and battery optimisation in settings page
2. Take backups
3. Reset both settings
4. Restore the backup
5. Check settings page: it shouldn't display correctly (it would say “optimising battery use” while it's not, would be fixed after a restart).

**Situation#4**
0. Take a fresh app
1. Enable background restriction and disable battery optimisation in settings page
2. Take backups
3. Reset both settings
4. Restore the backup
5. Check settings page: it shouldn't display correctly (it would say “optimising battery use” and “restricted” while they're not, would be fixed after a restart).

As you can see, the AOSP settings page cannot see the immediate changes to the battery optimisation setting.

**Comment by Cyberavater** [comment]
> As you can see, the AOSP settings page cannot see the immediate changes to the battery optimization setting.

Ahh, I found where we are out of sync. I fully agree that " the AOSP settings page cannot see the immediate changes to the battery optimization setting", but the thing is I'm not talking about the _battery optimization setting_ rather I'm talking about **Enable background restriction** which is a different setting.

Let me elaborate more just in case:

1. Let's change _battery optimization setting_ using AM, 
    Result- We will see that the AOSP settings page cannot see the immediate changes to the battery optimization setting.
    Conclusion- AOSP settings page is buggy when comes to reporting _battery optimization setting_ state. But! I don't care about that, this issue has nothing to do with that.

2. Let's change **Enable background restriction** using AM,
    Result- We will see the changes in the AOSP settings page **immediately**. Everything is working as expected.
3. Let the restore script/feature be the reason that enables the **Enable background restriction** (i.e. restore an app where **background restriction** was enabled).
    Result- We see the changes in the AOSP settings page **immediately**, but! it's stuck there and I can't change it back from the AOSP settings page and upon further inspection, I can even see that the app can actually run in the background which didn't happen at (2).
    Conclusion- From 2 and 3 (Ignoring 1 since (again!) this issue has nothing to do with _battery optimization setting_), Something is off with restoring feature as the mentioned issues only happen in 3 (again, these issues doesn't occur in 2)

Finally, I feel sorry (for taking so much time of the dev) that the dev wrote all that to explain it to me and really thankful that he did for me. I'm also doing my best to make it as clear as possible. If there is still confusion, feel free to ask in as short as possible (no need to waste so much of your time).

**Comment by MuntashirAkon** [comment]
> 3\. Let the restore script/feature be the reason that enables the **Enable background restriction** (i.e. restore an app where **background restriction** was enabled).
>     Result- We see the changes in the AOSP settings page **immediately**, but! it's stuck there and I can't change it back from the AOSP settings page

This is also a known behaviour of the AOSP settings. It can only handle `RUN_ANY_IN_BACKGROUND`, and not `RUN_IN_BACKGROUND`. If you disable both, the switch will simply stop working.

>  and upon further inspection, I can even see that the app can actually run in the background which didn't happen at (2).

The app operations won't work properly (it will take some time to take effect) unless you force-stop the app. Here's an workflow:
1. Open Android Terminal (available on F-Droid) and close it (a foreground service will still run in the background)
2. Disable background operation (from either the settings page or AM)
3. The service should be stopped but it doesn't
4. Force-stop the app and launch it again
5. Now, Close the app and the service should be closed immediately

**Comment by Cyberavater** [comment]
> This is also a known behaviour of the AOSP settings. It can only handle RUN_ANY_IN_BACKGROUND, and not RUN_IN_BACKGROUND. If you disable both, the switch will simply stop working.

So, If the user only disables RUN_ANY_IN_BACKGROUND, the restore script should not disable both? It should only disable both when the user backs up disabling both.

>The app operations won't work properly (it will take some time to take effect) unless you force-stop the app.

We can't do much for the AOSP, but should we not implement auto force-stop, when the user is using AM to do the job?

Thanks!

**Comment by MuntashirAkon** [comment]
> So, If the user only disables RUN_ANY_IN_BACKGROUND, the restore script should not disable both?

We don't explicitly check what users did or didn't. We simply back up/restore the current settings/permission without any prejudice. If something's wrong, it's most likely from the user-end.

However, if the app is currently installed, we only restore a setting/app op if that has been changed (i.e. we do not attempt to enable battery optimisation, we only disable it if it's present in the backups). This way current settings can also be preserved for installed apps.

**Comment by MuntashirAkon** [comment]
> should we not implement auto force-stop, when the user is using AM to do the job?

No, because it has other consequences (such as, some app ops can be reset if you force-stop an app). It's better to wait for a few minutes so that the operation can take effect.

**Comment by Cyberavater** [comment]
> We don't explicitly check what users did or didn't. We simply back up/restore the current settings/permission without any prejudice. If something's wrong, it's most likely from the user end.

> This is also a known behaviour of the AOSP settings. It can only handle RUN_ANY_IN_BACKGROUND, and not RUN_IN_BACKGROUND. If you disable both, the switch will simply stop working.

Then why would the settings be stuck if I restrict it using the AOSP settings page and then backup, uninstall and restore?

**Comment by MuntashirAkon** [comment]
> Then why would the settings be stuck if I restrict it using the AOSP settings page and then backup, uninstall and restore?

That's another bug (although it may seem similar to the above). The mode of an app op can be set for packages (removed in Android 11) and UIDs separately. AM and Rikka's App Ops set mode in both of these places, as opposed to the AOSP settings which only applies to the package. As a result, although the AOSP settings had correctly set the mode (to `allow`), it didn't work as the mode set for the UID has higher priority than the one set for the package.

In short, AOSP settings page is full of bugs, presumably because it hasn't been updated for quite some time.

**Comment by Cyberavater** [comment]
Again, thanks for your time and effort in explaining all these, really appreciate ya.

I've updated the OP accordingly.

**Comment by MuntashirAkon** [comment]
Should be fixed in 6c7a89af. Note that this will work only if your backups are recent since permission flags were not included in the old backups.
